sweep/sweeper.go

```diff
@@ -1,40 +1,701 @@
+	// list that starts with sets created from all inputs. However, there is
+	// a chance that those txes will not publish, because they already
+	// contain inputs that failed before. Therefore we also add sets
+	// consisting of only new inputs to the list, to make sure that new
+	// inputs are given a good, isolated chance of being published.
+	var newInputs, retryInputs []Input
+	for _, input := range s.pendingInputs {
+		// Skip inputs that have a minimum publish height that is not
+		// yet reached.
+		if input.minPublishHeight > currentHeight {
+			continue
+		}
+
+		// Add input to the either one of the lists.
+		if input.publishAttempts == 0 {
+			newInputs = append(newInputs, input.input)
+		} else {
+			retryInputs = append(retryInputs, input.input)
+		}
+	}
+
+	// If there is anything to retry, combine it with the new inputs and
+	// form input sets.
+	var allSets []inputSet
+	if len(retryInputs) > 0 {
+		var err error
+		allSets, err = generateInputPartitionings(
+			append(retryInputs, newInputs...),
+			s.relayFeePerKW, satPerKW,
+			s.cfg.MaxInputsPerTx,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create sets for just the new inputs.
+	newSets, err := generateInputPartitionings(
+		newInputs,
+		s.relayFeePerKW, satPerKW,
+		s.cfg.MaxInputsPerTx,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Append the new sets at the end of the list, because those tx likely
+	// have a higher fee per input.
+	return append(allSets, newSets...), nil
+}
+
+// sweep takes a set of preselected inputs, creates a sweep tx and publishes the
+// tx. The output address is only marked as used if the publish succeeds.
+func (s *UtxoSweeper) sweep(inputs inputSet,
+	satPerKW lnwallet.SatPerKWeight, currentHeight int32) error {
+
+	var err error
+
+	// Generate output script if no unused script available.
+	if s.currentOutputScript == nil {
+		s.currentOutputScript, err = s.cfg.GenSweepScript()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create sweep tx.
+	tx, err := createSweepTx(
+		inputs, s.currentOutputScript,
+		uint32(currentHeight), satPerKW, s.cfg.Signer,
+	)
+	if err != nil {
+		return fmt.Errorf("cannot create sweep tx: %v", err)
+	}
+
+	// Add tx before publication, so that we will always know that a spend
+	// by this tx is ours. Otherwise if the publish doesn't return, but did
+	// publish, we loose track of this tx. Even republication on startup
+	// doesn't prevent this, because that call returns a double spend error
+	// then and would also not add the hash to the store.
+	err = s.cfg.Store.NotifyPublishTx(tx)
+	if err != nil {
+		return fmt.Errorf("cannot notify publish: %v", err)
+	}
+
+	// Publish sweep tx.
+	log.Debugf("Publishing sweep tx %v", tx.TxHash())
+	err = s.cfg.PublishTransaction(tx)
+
+	// Keep outputScript in case of an error, so that it can be
+	// reused for the next tx and causes no address inflation.
+	if err != nil {
+		log.Error(err)
```

Comment:
Log should `"Failed to publish sweep tx: %v", err`

Reply:
Fixed in later commit.